Tests use accounts

glasserc opened the pull request:
Since Kinto/kinto#1736 landed, the test suite for both this library and kinto.js have been failing constantly (see e.g. #300, #301, #302, #303). Here is a fix which moves all tests to using accounts which should fix it.

Note that with this fix, the test suite takes a lot longer -- over 20 minutes on my laptop. I'm not exactly sure why but I bet it's due to the increased need for cryptographically hashing passwords once per request. Of special note is `should handle long list of changes`, which took almost 2 minutes on my machine. Maybe @leplatrem can make it faster on master, and maybe we should leave it on basicauth for the older kintos?

leplatrem commented on src/base.js:
```diff
@@ -684,4 +684,15 @@ export default class KintoClientBase {
   }
+
+  @capable(["accounts"])
+  async createAccount(username, password) {
```
We should mention this in the README

leplatrem commented on README.md:
```diff
@@ -163,7 +163,7 @@ const kinto = new KintoClient("https://my.server.tld/v1", {
 > #### Notes
 >
-> - As explained in the [server docs](http://kinto.readthedocs.io/en/stable/api/1.x/authentication.html#basic-auth), any string is accepted. You're not obliged to use the `username:password` format.
+> - As Kinto Accounts has become the default as of 10.2.0, you may need to create an account first.
```
Probably v11

leplatrem commented on src/requests.js:
```diff
@@ -30,8 +30,9 @@ export function createRequest(path, { data, permissions }, options = {}) {
+  const method = options.method || (data && data.id) ? "PUT" : "POST";
   return {
-    method: data && data.id ? "PUT" : "POST",
+    method: method,
```
method is superflous

leplatrem commented on test/kinto-8.1.2.ini:
```diff
@@ -24,6 +20,11 @@ kinto.attachment.base_path = /tmp
 
+# Force accounts (not default until 10.2)
+multiauth.policies = account
+multiauth.policy.account.use = kinto.plugins.accounts.authentication.AccountsAuthenticationPolicy
```
`kinto.plugins.accounts.AccountsPolicy`

leplatrem commented on .travis.yml:
```diff
@@ -5,7 +5,7 @@ env:
 - ACTION="run dist" # ensures building dist files doesn't break
 - ACTION="run cs-check"
-- SERVER=6.0.1 ACTION=test
+- SERVER=8.1.2 ACTION=test
```
some old version on purpose?

glasserc replied:
I don't understand. 8.1.2 *is* the old version on purpose. 7.0.0 is the oldest one that supports `accounts`, but I didn't see a reason to keep anything from that era around.

glasserc commented:
Refs also Kinto/kinto#1821.

glasserc commented:
Rebased onto current master and cleaned up a little bit since #312 landed. This PR now serves to move tests to accounts and stop using basicauth in tests. This will still make running the test suite much more expensive, but fixing Kinto/kinto#1821 will have the same effect so maybe we should just go for it to ensure that we're testing a scenario that matches the Kinto "default".

leplatrem commented on README.md:
```diff
@@ -165,7 +165,7 @@ const kinto = new KintoClient("https://my.server.tld/v1", {
 > #### Notes
 >
-> - As explained in the [server docs](http://kinto.readthedocs.io/en/stable/api/1.x/authentication.html#basic-auth), any string is accepted. You're not obliged to use the `username:password` format.
+> - As Kinto Accounts has become the default as of 11, you may need to create an account first.
```
*as of version 11* ?

itaisteinherz commented on test/integration_test.js:
```diff
@@ -87,7 +87,9 @@ describe("Integration tests", function() {
 
-    beforeEach(() => server.flush());
+    beforeEach(() => {
+      return server.flush().then(_ => api.createAccount("user", "pass"));
```
```suggestion
      return server.flush().then(() => api.createAccount("user", "pass"));
```

itaisteinherz commented on test/integration_test.js:
```diff
@@ -319,6 +321,8 @@ describe("Integration tests", function() {
               data = data.filter(p => !isBucketCreate(p));
             }
+            const isOwnAccount = p => p.uri == "/accounts/user";
+            data = data.filter(p => !isOwnAccount(p));
```
Would be simpler to remove the line above and change this to:

```suggestion
            data = data.filter(p => p.uri !== "/accounts/user");
```

Also, `p` is not a very clear variable name, and so it would be preferable to use a longer and more readble one.

itaisteinherz commented on test/integration_test.js:
```diff
@@ -575,7 +580,9 @@ describe("Integration tests", function() {
 
-    beforeEach(() => server.flush());
+    beforeEach(() => {
+      return server.flush().then(_ => api.createAccount("user", "pass"));
```
```suggestion
      return server.flush().then(() => api.createAccount("user", "pass"));
```

(Same as https://github.com/Kinto/kinto-http.js/pull/304/files#r249295696)

itaisteinherz commented on test/integration_test.js:
```diff
@@ -646,7 +651,9 @@ describe("Integration tests", function() {
 
-    beforeEach(() => server.flush());
+    beforeEach(() => {
+      return server.flush().then(_ => api.createAccount("user", "pass"));
```
```suggestion
      return server.flush().then(() => api.createAccount("user", "pass"));
```

(Same as https://github.com/Kinto/kinto-http.js/pull/304/files#r249295696)

itaisteinherz commented on test/integration_test.js:
```diff
@@ -680,7 +687,9 @@ describe("Integration tests", function() {
 
-    beforeEach(() => server.flush());
+    beforeEach(() => {
+      return server.flush().then(_ => api.createAccount("user", "pass"));
```
```suggestion
      return server.flush().then(() => api.createAccount("user", "pass"));
```

(Same as https://github.com/Kinto/kinto-http.js/pull/304/files#r249295696)

leplatrem commented:
Closing. Reopen if willing to see this landing :)
